pkg/kubernetes/reconcile.go

```diff
@@ -67,64 +70,73 @@ func extract(deep interface{}) (map[string]manifest.Manifest, error) {
 	return extracted, nil
 }
 
-// walkJSON traverses deeply nested kubernetes manifest and extracts them into a flat []dict.
-func walkJSON(deep interface{}, extracted map[string]manifest.Manifest, path trace) error {
-	// array: walkJSON for each
-	if d, ok := deep.([]map[string]interface{}); ok {
-		for i, j := range d {
-			path := append(path, fmt.Sprintf("[%v]", i))
-			if err := walkJSON(j, extracted, path); err != nil {
-				return err
+// walkJSON recurses into either a map or list, returning a list of all objects that look
+// like kubernetes resources. We support resources at an arbitrary level of nesting, and
+// return an error if a node is not walkable.
+//
+// Handling the different types is quite gross, so we split this method into a generic
+// walkJSON, and then walkObj/walkList to handle the two different types of collection we
+// support.
+func walkJSON(ptr interface{}, extracted map[string]manifest.Manifest, path trace) error {
+	switch v := ptr.(type) {
+	case map[string]interface{}:
+		return walkObj(v, extracted, path)
+	case []map[string]interface{}:
+		return walkList(v, extracted, path)
+	case []interface{}:
+		l := make([]map[string]interface{}, 0, len(v))
+		for _, o := range v {
+			if dict, ok := o.(map[string]interface{}); ok {
```

Comment:
This assumes every `[]interface{}` always contains a `map[string]interface{}`, otherwise the value is discarded.

While this is fine in most cases, it would miss lists nested in lists ... something like:

```jsonnet
[
  [
    { /* ... */ }
  ]
]
```

would not work. I guess such patterns can happen in real world use, so this might be worth adding here